File: controlplane/internal/authz/authz_test.go
package authz

import "testing"

// Port of AuthzTest.kt — 278 LOC, 14 cases, unit (in-memory CedarEngine), THE SEED ORACLE.
// 02-authz.md §10. Case names are verbatim from the Kotlin.

// AuthzTest.kt:28-34 — the seed policy set.
var authzTestSeedPolicies = map[int64]string{
	1: `permit(principal in Role::"system:admin", action in [Action::"admin.datasources",Action::"admin.policies",Action::"admin.identity"], resource);`,
	2: `forbid(principal, action == Action::"task.approve", resource) when { principal == resource.requester };`,
	3: `permit(principal in Role::"system:admin", action == Action::"task.approve", resource);`,
	4: `permit(principal, action == Action::"audit.read", resource) when { resource is AuditRecord && resource.principal == principal };`,
	5: `permit(principal in Role::"auditor", action == Action::"audit.read", resource);`,
}

// AuthzTest.kt:145-154 — a role-scoped approval grant.
var authzTestRoleScopedApproval = map[int64]string{
	1: `permit(principal in Role::"pii-reader", action == Action::"task.approve", resource) when { resource in Role::"pii-reader" };`,
}

// 1. system-admin is allowed on admin actions
func TestAuthz_SystemAdminIsAllowedOnAdminActions(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{"root": {"system:admin"}})
	for _, action := range []AuthzAction{ActionAdminPolicies, ActionAdminDatasources, ActionAdminIdentity} {
		assertAllow(t, a.Authorize("root", action, ResourceSystem{}, AuthzContext{}), string(action))
	}
}

// 2. 🔒 no roles is denied on admin actions — the "admin = any session" hole stays closed
func TestAuthz_NoRolesIsDeniedOnAdminActions(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{})
	d := a.Authorize("nobody", ActionAdminPolicies, ResourceSystem{}, AuthzContext{})
	assertDeny(t, d, "admin.policies with no roles")
	// The Kotlin oracle records the reason as branch 3: reasons are empty, so it is the
	// deny-by-default message rather than "denied by policy: …".
	if d.Reason != "no policy permits this action" {
		t.Errorf("reason = %q, want %q", d.Reason, "no policy permits this action")
	}
	if d.Code != "forbidden" {
		t.Errorf("code = %q, want %q", d.Code, "forbidden")
	}
}

// 3. a role other than system-admin is still denied on admin actions
func TestAuthz_ANonAdminRoleIsStillDeniedOnAdminActions(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{"analyst-alice": {"analyst"}})
	assertDeny(t, a.Authorize("analyst-alice", ActionAdminPolicies, ResourceSystem{}, AuthzContext{}), "analyst on admin.policies")
}

// 4. the audit policies allow own records and grant auditors the whole collection
//
// Pins the `resource is AuditRecord && resource.principal == principal` type-guard + narrowing shape,
// and AuditLog::"all" as a distinct type.
func TestAuthz_AuditPoliciesAllowOwnRecordsAndGrantAuditorsTheWholeCollection(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{
		"alice": {"analyst"},
		"bob":   {"auditor"},
	})
	assertAllow(t, a.Authorize("alice", ActionAuditRead, ResourceAuditRecord{Principal: "alice"}, AuthzContext{}), "alice on her own record")
	assertDeny(t, a.Authorize("alice", ActionAuditRead, ResourceAuditRecord{Principal: "bob"}, AuthzContext{}), "alice on bob's record")
	assertDeny(t, a.Authorize("alice", ActionAuditRead, ResourceAuditLog{}, AuthzContext{}), "alice on the whole log")
	assertAllow(t, a.Authorize("bob", ActionAuditRead, ResourceAuditRecord{Principal: "alice"}, AuthzContext{}), "auditor on alice's record")
	assertAllow(t, a.Authorize("bob", ActionAuditRead, ResourceAuditLog{}, AuthzContext{}), "auditor on the whole log")
}

// 5. an approver may approve someone else's request
//
// EUID is Request::"<requester>#<datasourceName ?: "-">" -> Request::"someone-else#-".
func TestAuthz_AnApproverMayApproveSomeoneElsesRequest(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{"approver1": {"system:admin"}})
	d := a.Authorize("approver1", ActionTaskApprove,
		ResourceApprovalRequest{Requester: "someone-else"}, AuthzContext{})
	assertAllow(t, d, "system-admin approving someone else")
}

// 6. 🔒 ROLE self-approval is denied even for a system-admin — the self-approval hole stays closed
//
// Pins Cedar's forbid-overrides-permit precedence: forbid id 2 beats permit id 3.
//
// This is ALSO the case the errors-first mapping exists to protect (98-cedar-spike-report.md § S4): with
// the Request entity omitted or its `requester` attribute missing, the forbid ERRORS, cedar-go drops it,
// and the permit stands — a verdict-first port would Allow here. See errors_first_test.go.
func TestAuthz_RoleSelfApprovalIsDeniedEvenForASystemAdmin(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{"alice": {"system:admin"}})
	d := a.Authorize("alice", ActionTaskApprove,
		ResourceApprovalRequest{Requester: "alice"}, AuthzContext{})
	assertDeny(t, d, "system-admin self-approval")
}

// 7. 🔒 a non-admin approving someone else's request is still denied — no ambient permit
func TestAuthz_ANonAdminApprovingSomeoneElsesRequestIsStillDenied(t *testing.T) {
	a := authzFor(t, authzTestSeedPolicies, map[string][]string{"bob": {}})
	assertDeny(t, a.Authorize("bob", ActionTaskApprove,
		ResourceApprovalRequest{Requester: "someone-else"}, AuthzContext{}), "non-admin approving")
}

// 8. an approval request scoped to a role is matched via `resource in Role` — Request carries the role
// as a Cedar parent
//
// S6 KEY CASE — this is the dedupeByEuid collision: the Request's Role parent (Role::"pii-reader") is
// the SAME EUID the principal already carries as a parent, so the entity list holds two distinct Entity
// objects for one EntityUID. cedar-java rejects that outright; the test asserts ALLOW, i.e. that
// dedupeByEuid PREVENTED the error. It does NOT assert the error itself.
func TestAuthz_AnApprovalRequestScopedToARoleIsMatchedViaResourceInRole(t *testing.T) {
	a := authzFor(t, authzTestRoleScopedApproval, map[string][]string{"approver1": {"pii-reader"}})
	d := a.Authorize("approver1", ActionTaskApprove, ResourceApprovalRequest{
		Requester: "someone-else",
		RoleName:  ptr("pii-reader"),
	}, AuthzContext{})
	assertAllow(t, d, "role-scoped approval matching the principal's own role")
}

// 9. a request scoped to a DIFFERENT role than the policy grants is denied — the Role parent is the
// request's, not ambient
func TestAuthz_ARequestScopedToADifferentRoleIsDenied(t *testing.T) {
	a := authzFor(t, authzTestRoleScopedApproval, map[string][]string{"approver1": {"pii-reader"}})
	d := a.Authorize("approver1", ActionTaskApprove, ResourceApprovalRequest{
		Requester: "someone-else",
		RoleName:  ptr("other-role"),
	}, AuthzContext{})
	assertDeny(t, d, "request scoped to other-role")
}

// 10. task lifecycle and grant revoke actions validate against the bundled schema
func TestAuthz_TaskLifecycleAndGrantRevokeActionsValidateAgainstTheBundledSchema(t *testing.T) {
	for _, src := range []string{
		`permit(principal, action == Action::"task.request", resource);`,
		`permit(principal, action == Action::"task.approve", resource);`,
		`permit(principal, action == Action::"task.read", resource);`,
		`permit(principal, action == Action::"task.assume", resource);`,
		`permit(principal, action == Action::"task.cancel", resource);`,
		`permit(principal, action == Action::"task.delete", resource);`,
		`permit(principal, action == Action::"grant.revoke", resource);`,
	} {
		assertValid(t, src)
	}
}

// 11. task assume seeds validate and allow only parties or auditor
//
// Pins the `resource has approver` guard and that an admin's task.read does NOT confer task.assume.
func TestAuthz_TaskAssumeSeedsValidateAndAllowOnlyPartiesOrAuditor(t *testing.T) {
	policies := map[int64]string{
		21: `permit(principal, action == Action::"task.assume", resource) when { resource is Request && (resource.requester == principal || (resource has approver && resource.approver == principal)) };`,
		22: `permit(principal in Role::"system:auditor", action == Action::"task.assume", resource) when { resource is Request };`,
		23: `permit(principal in Role::"system:admin", action == Action::"task.read", resource);`,
	}
	assertValid(t, policies[21])
	assertValid(t, policies[22])

	a := authzFor(t, policies, map[string][]string{
		"auditor@example.com": {"system:auditor"},
		"admin@example.com":   {"system:admin"},
	})
	resource := ResourceApprovalRequest{
		Requester:  "requester@example.com",
		Approver:   ptr("approver@example.com"),
		ExecutedBy: ptr("executor@example.com"),
	}
	assertAllow(t, a.Authorize("requester@example.com", ActionTaskAssume, resource, AuthzContext{}), "requester")
	assertAllow(t, a.Authorize("approver@example.com", ActionTaskAssume, resource, AuthzContext{}), "approver")
	assertAllow(t, a.Authorize("auditor@example.com", ActionTaskAssume, resource, AuthzContext{}), "auditor")
	assertDeny(t, a.Authorize("admin@example.com", ActionTaskAssume, resource, AuthzContext{}),
		"admin — task.read is not task.assume")
}

// 12. approval request marshals `executedBy` as a User attribute
//
// approver ABSENT here, executedBy present — pins the optional-attribute marshalling shape (INV-A2-3's
// sibling on Request).
func TestAuthz_ApprovalRequestMarshalsExecutedByAsAUserAttribute(t *testing.T) {
	a := authzFor(t, map[int64]string{
		1: `permit(principal, action == Action::"task.assume", resource) when { resource has executedBy && resource.executedBy == principal };`,
	}, map[string][]string{})
	d := a.Authorize("executor@example.com", ActionTaskAssume, ResourceApprovalRequest{
		Requester:  "requester@example.com",
		ExecutedBy: ptr("executor@example.com"),
	}, AuthzContext{})
	assertAllow(t, d, "executedBy match")
}

// 13. retired workflow action ids are rejected by the bundled schema
//
// S1/S2 KEY CASE — the ONLY place cedar-java's REJECT of an undeclared action is pinned at the
// CedarSchema.validate() level. The assertion is only isNotEmpty(): the error MESSAGE TEXT is never
// pinned, so cedar-go's wording is free to differ.
func TestAuthz_RetiredWorkflowActionIDsAreRejectedByTheBundledSchema(t *testing.T) {
	for _, src := range []string{
		`permit(principal, action == Action::"workflow.request", resource);`,
		`permit(principal, action == Action::"workflow.approve", resource);`,
		`permit(principal, action == Action::"workflow.read", resource);`,
		`permit(principal, action == Action::"workflow.readResult", resource);`,
		`permit(principal, action == Action::"workflow.revoke", resource);`,
	} {
		assertInvalid(t, src)
	}
}

// 14. an approval request scoped to a datasource is matched via `resource in Datasource` by NAME
//
// INV-A2-2: the Request EUID embeds the datasource NAME and the Datasource parent is NAME-keyed.
func TestAuthz_AnApprovalRequestScopedToADatasourceIsMatchedByName(t *testing.T) {
	a := authzFor(t, map[int64]string{
		1: `permit(principal in Role::"acme-approver", action == Action::"task.approve", resource) when { resource in Datasource::"acme-mysql" };`,
	}, map[string][]string{"approver1": {"acme-approver"}})

	assertAllow(t, a.Authorize("approver1", ActionTaskApprove, ResourceApprovalRequest{
		Requester:      "someone-else",
		DatasourceName: ptr("acme-mysql"),
	}, AuthzContext{}), "scoped to acme-mysql")

	assertDeny(t, a.Authorize("approver1", ActionTaskApprove, ResourceApprovalRequest{
		Requester:      "someone-else",
		DatasourceName: ptr("some-other-ds"),
	}, AuthzContext{}), "scoped to some-other-ds")
}

File: control-plane/src/test/kotlin/com/ridi/oss/proxymonster/controlplane/AuthzTest.kt
package com.ridi.oss.proxymonster.controlplane

import com.ridi.oss.proxymonster.controlplane.authz.Authz
import com.ridi.oss.proxymonster.controlplane.authz.AuthzAction
import com.ridi.oss.proxymonster.controlplane.authz.AuthzDecision
import com.ridi.oss.proxymonster.controlplane.authz.AuthzResource
import com.ridi.oss.proxymonster.controlplane.authz.CedarEngine
import com.ridi.oss.proxymonster.controlplane.authz.CedarPolicyStore
import com.ridi.oss.proxymonster.controlplane.authz.CedarSchema
import com.ridi.oss.proxymonster.controlplane.authz.RoleSource
import java.sql.Connection
import java.util.logging.Logger
import javax.sql.DataSource
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertTrue

/**
 * Pure, no-DB proof that the seeded admin-gating and audit decisions hold at the Cedar-decision level: admin
 * routes require a real `system:admin` role (not "any session"), ROLE self-approval is denied even for an
 * admin, ordinary principals can read only their own audit records, and auditors can read the whole
 * collection. [CedarEngine]'s in-memory constructor and [Authz]'s unused [CedarPolicyStore] parameter
 * (a DataSource that's never actually queried — `authorize()` never touches it, see Authz.kt) keep
 * this fully off any JDBC/Docker dependency.
 */
class AuthzTest {
    private val seedPolicies = listOf(
        1L to """permit(principal in Role::"system:admin", action in [Action::"admin.datasources",Action::"admin.policies",Action::"admin.identity"], resource);""",
        2L to """forbid(principal, action == Action::"task.approve", resource) when { principal == resource.requester };""",
        3L to """permit(principal in Role::"system:admin", action == Action::"task.approve", resource);""",
        4L to """permit(principal, action == Action::"audit.read", resource) when { resource is AuditRecord && resource.principal == principal };""",
        5L to """permit(principal in Role::"auditor", action == Action::"audit.read", resource);""",
    )

    /** A [DataSource] that's never actually connected to — [Authz] never touches its [CedarPolicyStore]
     *  parameter, only [CedarEngine] does, and this test's engine is built from an in-memory policy list. */
    private object UnusedDataSource : DataSource {
        override fun getConnection(): Connection = error("not used by this test")
        override fun getConnection(username: String?, password: String?): Connection = error("not used by this test")
        override fun getLogWriter() = error("not used by this test")
        override fun setLogWriter(out: java.io.PrintWriter?) = error("not used by this test")
        override fun setLoginTimeout(seconds: Int) = error("not used by this test")
        override fun getLoginTimeout() = error("not used by this test")
        override fun getParentLogger(): Logger = error("not used by this test")
        override fun <T : Any?> unwrap(iface: Class<T>?): T = error("not used by this test")
        override fun isWrapperFor(iface: Class<*>?): Boolean = false
    }

    private fun authzWithRoles(roles: Map<String, Set<String>>): Authz {
        val engine = CedarEngine(seedPolicies)
        val policyStore = CedarPolicyStore(UnusedDataSource)
        val roleSource = RoleSource { principal -> roles[principal] ?: emptySet() }
        return Authz(engine, policyStore, roleSource)
    }

    @Test
    fun `system-admin is allowed on admin actions`() {
        val authz = authzWithRoles(mapOf("root" to setOf("system:admin")))
        assertEquals(AuthzDecision.Allow, authz.authorize("root", AuthzAction.ADMIN_POLICIES, AuthzResource.System))
        assertEquals(AuthzDecision.Allow, authz.authorize("root", AuthzAction.ADMIN_DATASOURCES, AuthzResource.System))
        assertEquals(AuthzDecision.Allow, authz.authorize("root", AuthzAction.ADMIN_IDENTITY, AuthzResource.System))
    }

    @Test
    fun `no roles is denied on admin actions — the 'admin = any session' hole stays closed`() {
        val authz = authzWithRoles(emptyMap())
        val decision = authz.authorize("nobody", AuthzAction.ADMIN_POLICIES, AuthzResource.System)
        assertIs<AuthzDecision.Deny>(decision)
    }

    @Test
    fun `a role other than system-admin is still denied on admin actions`() {
        val authz = authzWithRoles(mapOf("analyst-alice" to setOf("analyst")))
        val decision = authz.authorize("analyst-alice", AuthzAction.ADMIN_POLICIES, AuthzResource.System)
        assertIs<AuthzDecision.Deny>(decision)
    }

    @Test
    fun `the audit policies allow own records and grant auditors the whole collection`() {
        val authz = authzWithRoles(
            mapOf(
                "alice" to setOf("analyst"),
                "bob" to setOf("auditor"),
            ),
        )

        assertEquals(
            AuthzDecision.Allow,
            authz.authorize("alice", AuthzAction.AUDIT_READ, AuthzResource.AuditRecord(principal = "alice")),
        )
        assertIs<AuthzDecision.Deny>(
            authz.authorize("alice", AuthzAction.AUDIT_READ, AuthzResource.AuditRecord(principal = "bob")),
        )
        assertIs<AuthzDecision.Deny>(
            authz.authorize("alice", AuthzAction.AUDIT_READ, AuthzResource.AuditLog),
        )
        assertEquals(
            AuthzDecision.Allow,
            authz.authorize("bob", AuthzAction.AUDIT_READ, AuthzResource.AuditRecord(principal = "alice")),
        )
        assertEquals(
            AuthzDecision.Allow,
            authz.authorize("bob", AuthzAction.AUDIT_READ, AuthzResource.AuditLog),
        )
    }

    @Test
    fun `an approver may approve someone else's request`() {
        val authz = authzWithRoles(mapOf("approver1" to setOf("system:admin")))
        val decision = authz.authorize(
            "approver1",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "someone-else"),
        )
        assertEquals(AuthzDecision.Allow, decision)
    }

    @Test
    fun `ROLE self-approval is denied even for a system-admin — the self-approval hole stays closed`() {
        val authz = authzWithRoles(mapOf("alice" to setOf("system:admin")))
        val decision = authz.authorize(
            "alice",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "alice"),
        )
        assertIs<AuthzDecision.Deny>(decision)
    }

    @Test
    fun `a non-admin approving someone else's request is still denied — no ambient permit`() {
        val authz = authzWithRoles(mapOf("bob" to emptySet()))
        val decision = authz.authorize(
            "bob",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "someone-else"),
        )
        assertIs<AuthzDecision.Deny>(decision)
    }

    /** A Request scoped to a role (`ApprovalRequest.roleName`) attaches that Role as a Cedar PARENT
     *  (`Request in [Datasource, Role]`, schema.cedarschema), so `resource in Role::"..."` is a real
     *  entity-membership check — not a string comparison — letting a policy scope who may approve by
     *  the ROLE being requested, not just the requester's identity. */
    private fun authzForRoleScopedApproval(): Authz {
        val engine = CedarEngine(
            listOf(
                1L to """permit(principal in Role::"pii-reader", action == Action::"task.approve", resource) when { resource in Role::"pii-reader" };""",
            ),
        )
        val policyStore = CedarPolicyStore(UnusedDataSource)
        val roleSource = RoleSource { principal -> if (principal == "approver1") setOf("pii-reader") else emptySet() }
        return Authz(engine, policyStore, roleSource)
    }

    @Test
    fun `an approval request scoped to a role is matched via 'resource in Role' — Request carries the role as a Cedar parent`() {
        val authz = authzForRoleScopedApproval()
        val decision = authz.authorize(
            "approver1",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "someone-else", roleName = "pii-reader"),
        )
        assertEquals(AuthzDecision.Allow, decision)
    }

    @Test
    fun `a request scoped to a DIFFERENT role than the policy grants is denied — the Role parent is the request's, not ambient`() {
        val authz = authzForRoleScopedApproval()
        val decision = authz.authorize(
            "approver1",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "someone-else", roleName = "other-role"),
        )
        assertIs<AuthzDecision.Deny>(decision)
    }

    @Test
    fun `a request-specific policy cannot carry over to a later request with the same requester and datasource`() {
        // An operator permit scoped to one exact Request EUID must match only that persisted request, not a
        // later one that happens to share the requester and datasource — each durable request keys off its id.
        // Routed through `toApprovalResource`, the single persisted-path constructor every route funnels
        // through, so dropping the id there (the exact regression this closes) would fail here.
        val authz = Authz(
            CedarEngine(listOf(1L to """permit(principal, action == Action::"task.approve", resource == Request::"41");""")),
            CedarPolicyStore(UnusedDataSource),
            RoleSource { emptySet() },
        )
        fun row(id: Long) = AccessRequest(
            id = id, principal = "requester@example.com", datasourceName = "acme",
            requestedDurationSec = 3600, status = "PENDING", createdAt = "2026-01-01T00:00:00Z",
        )
        assertEquals(
            AuthzDecision.Allow,
            authz.authorize("approver@example.com", AuthzAction.TASK_APPROVE, row(41).toApprovalResource()),
        )
        assertIs<AuthzDecision.Deny>(
            authz.authorize("approver@example.com", AuthzAction.TASK_APPROVE, row(42).toApprovalResource()),
        )
    }

    /** Regression: the Request's Datasource parent must be keyed by NAME (`Datasource::"acme-mysql"`),
     *  matching every other Datasource-keyed grant in the system — including this exact policy shape,
     *  straight from the doc's worked examples. Keyed by numeric id (`Datasource::"2"`) instead, a
     *  per-datasource approval policy silently never matches. */
    @Test
    fun `task lifecycle and grant revoke actions validate against the bundled schema`() {
        val sources = listOf(
            """permit(principal, action == Action::"task.request", resource);""",
            """permit(principal, action == Action::"task.approve", resource);""",
            """permit(principal, action == Action::"task.read", resource);""",
            """permit(principal, action == Action::"task.assume", resource);""",
            """permit(principal, action == Action::"task.cancel", resource);""",
            """permit(principal, action == Action::"task.delete", resource);""",
            """permit(principal, action == Action::"grant.revoke", resource);""",
        )
        sources.forEach { source -> assertTrue(CedarSchema.validate(source).isEmpty(), source) }
    }

    @Test
    fun `task assume seeds validate and allow only parties or auditor`() {
        val parties = """permit(principal, action == Action::"task.assume", resource) when { resource is Request && (resource.requester == principal || (resource has approver && resource.approver == principal)) };"""
        val auditor = """permit(principal in Role::"system:auditor", action == Action::"task.assume", resource) when { resource is Request };"""
        assertTrue(CedarSchema.validate(parties).isEmpty(), parties)
        assertTrue(CedarSchema.validate(auditor).isEmpty(), auditor)
        val engine = CedarEngine(listOf(21L to parties, 22L to auditor, 23L to """permit(principal in Role::"system:admin", action == Action::"task.read", resource);"""))
        val authz = Authz(
            engine,
            CedarPolicyStore(UnusedDataSource),
            RoleSource { principal ->
                when (principal) {
                    "auditor@example.com" -> setOf("system:auditor")
                    "admin@example.com" -> setOf("system:admin")
                    else -> emptySet()
                }
            },
        )
        val request = AuthzResource.ApprovalRequest(
            requester = "requester@example.com",
            approver = "approver@example.com",
            executedBy = "executor@example.com",
        )
        assertEquals(AuthzDecision.Allow, authz.authorize("requester@example.com", AuthzAction.TASK_ASSUME, request))
        assertEquals(AuthzDecision.Allow, authz.authorize("approver@example.com", AuthzAction.TASK_ASSUME, request))
        assertEquals(AuthzDecision.Allow, authz.authorize("auditor@example.com", AuthzAction.TASK_ASSUME, request))
        assertIs<AuthzDecision.Deny>(authz.authorize("admin@example.com", AuthzAction.TASK_ASSUME, request))
    }

    @Test
    fun `approval request marshals executedBy as a User attribute`() {
        val source = """permit(principal, action == Action::"task.assume", resource) when { resource has executedBy && resource.executedBy == principal };"""
        val authz = Authz(CedarEngine(listOf(1L to source)), CedarPolicyStore(UnusedDataSource), RoleSource { emptySet() })
        assertEquals(
            AuthzDecision.Allow,
            authz.authorize(
                "executor@example.com",
                AuthzAction.TASK_ASSUME,
                AuthzResource.ApprovalRequest(requester = "requester@example.com", executedBy = "executor@example.com"),
            ),
        )
    }

    @Test
    fun `retired workflow action ids are rejected by the bundled schema`() {
        listOf(
            "workflow.request", "workflow.approve", "workflow.read", "workflow.readResult", "workflow.revoke",
        ).forEach { action ->
            assertTrue(
                CedarSchema.validate("""permit(principal, action == Action::"$action", resource);""").isNotEmpty(),
                action,
            )
        }
    }

    @Test
    fun `an approval request scoped to a datasource is matched via 'resource in Datasource' by NAME`() {
        val engine = CedarEngine(
            listOf(
                1L to """permit(principal in Role::"acme-approver", action == Action::"task.approve", resource) when { resource in Datasource::"acme-mysql" };""",
            ),
        )
        val policyStore = CedarPolicyStore(UnusedDataSource)
        val roleSource = RoleSource { principal -> if (principal == "approver1") setOf("acme-approver") else emptySet() }
        val authz = Authz(engine, policyStore, roleSource)

        val matching = authz.authorize(
            "approver1",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "someone-else", datasourceName = "acme-mysql"),
        )
        assertEquals(AuthzDecision.Allow, matching)

        // A request for a DIFFERENT datasource must not match — the Datasource parent is the
        // request's, not ambient.
        val other = authz.authorize(
            "approver1",
            AuthzAction.TASK_APPROVE,
            AuthzResource.ApprovalRequest(requester = "someone-else", datasourceName = "some-other-ds"),
        )
        assertIs<AuthzDecision.Deny>(other)
    }
}
